platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java

```diff
@@ -35,20 +40,47 @@ public class DiagnosticToProblemListener implements DiagnosticListener<JavaFileO
 
     private final ProblemReporter problemReporter;
 
+    /** A collection of filters that will serve as a predicate to determine if a diagnostic should be reported or not.*/
+    private final Collection<Predicate<Diagnostic<? extends JavaFileObject>>> diagnosticFilters = new ArrayList<>();
+
+    /**
+     * A map of explicit overrides between {@link Diagnostic.Kind} and {@link Severity}.
+     * <p>
+     * This will take precedence over the default mapping defined in {@link #mapKindToSeverity(Diagnostic.Kind)}.
+     */
+    private final Map<Diagnostic.Kind, Diagnostic.Kind> severityOverrides = new HashMap<>();
+
+
     public DiagnosticToProblemListener(ProblemReporter problemReporter) {
         this.problemReporter = problemReporter;
     }
 
+    public void addDiagnosticFilter(Predicate<Diagnostic<? extends JavaFileObject>> diagnosticFilter) {
+        diagnosticFilters.add(diagnosticFilter);
+    }
+
+    public void addSeverityOverride(Diagnostic.Kind fromKind, Diagnostic.Kind toKind) {
+        severityOverrides.put(fromKind, toKind);
+    }
+
     @Override
     public void report(Diagnostic<? extends JavaFileObject> diagnostic) {
+        // If any of the filters return true, we should not report the diagnostic
+        if (diagnosticFilters.stream().anyMatch(filter -> filter.test(diagnostic))) {
+            return;
```

Comment:
💅 Should this `report` method be renamed `onProblem` or `maybeReport` or something, as reporting now depends on a diagnostic surviving the filters?

Reply:
Meanwhile your reasoning is sound, I cannot as this method comes from [`DiagnosticListener`](https://github.com/openjdk/jdk/blob/master/src/java.compiler/share/classes/javax/tools/DiagnosticListener.java)